packages/back-end/src/jobs/addSafeRolloutSnapshotJob.ts

```diff
@@ -0,0 +1,99 @@
+import Agenda, { Job } from "agenda";
+import { getContextForAgendaJobByOrgId } from "back-end/src/services/organizations";
+import { logger } from "back-end/src/util/logger";
+import { getCollection } from "back-end/src/util/mongo.util";
+import { getFeature } from "back-end/src/models/FeatureModel";
+import { getSafeRolloutRuleFromFeature } from "back-end/src/routers/safe-rollout/safe-rollout.helper";
+import { createSafeRolloutSnapshot } from "back-end/src/services/safeRolloutSnapshots";
+import { trackJob } from "back-end/src/services/tracing";
+import { COLLECTION_NAME } from "back-end/src/models/SafeRolloutModel";
+import { SafeRolloutInterface } from "back-end/src/validators/safe-rollout";
+
+const UPDATE_SINGLE_SAFE_ROLLOUT_SNAPSHOT = "updateSingleSafeRolloutSnapshot";
+const QUEUE_SAFE_ROLLOUT_SNAPSHOT_UPDATES = "queueSafeRolloutSnapshotUpdates";
+
+type UpdateSingleSafeRolloutSnapshotJob = Job<{
+  safeRollout: SafeRolloutInterface;
+}>;
+
+export default async function (agenda: Agenda) {
+  agenda.define(QUEUE_SAFE_ROLLOUT_SNAPSHOT_UPDATES, async () => {
+    const safeRollouts = await getAllSafeRolloutsToUpdate();
+
+    for (const safeRollout of safeRollouts) {
+      await queueSafeRolloutSnapshotUpdate(safeRollout);
+    }
+  });
+
+  agenda.define(
+    UPDATE_SINGLE_SAFE_ROLLOUT_SNAPSHOT,
+    // This job queries a datasource, which may be slow. Give it 30 minutes to complete.
+    { lockLifetime: 30 * 60 * 1000 }, // 30 minutes
+    updateSingleSafeRolloutSnapshot
+  );
+
+  await startUpdateJob();
+
+  async function startUpdateJob() {
+    const updateResultsJob = agenda.create(
+      QUEUE_SAFE_ROLLOUT_SNAPSHOT_UPDATES,
+      {}
+    );
+    updateResultsJob.unique({});
+    updateResultsJob.repeatEvery("10 minutes");
+    await updateResultsJob.save();
+  }
+
+  async function queueSafeRolloutSnapshotUpdate(
+    safeRollout: SafeRolloutInterface
+  ) {
+    const job = agenda.create(UPDATE_SINGLE_SAFE_ROLLOUT_SNAPSHOT, {
+      safeRollout,
+    });
+    job.unique({ id: safeRollout.id });
+    job.schedule(new Date());
+    await job.save();
+  }
+}
+
+const updateSingleSafeRolloutSnapshot = trackJob(
+  UPDATE_SINGLE_SAFE_ROLLOUT_SNAPSHOT,
+  async (job: UpdateSingleSafeRolloutSnapshotJob) => {
+    const { safeRollout } = job.attrs.data;
+
+    const { ruleId, organization, featureId } = safeRollout;
+    if (!ruleId || !organization || !featureId) return;
+
+    const context = await getContextForAgendaJobByOrgId(organization);
+    const feature = await getFeature(context, featureId);
+    if (!feature) return;
+
+    const safeRolloutRule = getSafeRolloutRuleFromFeature(feature, ruleId);
+    if (!safeRolloutRule) return;
```

Comment:
We need a whole bunch of checks in here to skip snapshots if:
- The rule's environment toggle is off
- The feature is archived
- The rule itself is disabled

We also probably want some mechanism to clean up these orphaned rollout interfaces.  Otherwise, the list of rollouts to queue will continue growing indefinitely and this job will become very expensive and slow.